watertap/core/plugins/solvers.py

```diff
@@ -0,0 +1,150 @@
+                    min_scale=min_scale)
+        except AssertionError as err:
+            if str(err) == "Error in AMPL evaluation":
+                print("ipopt-watertap: Issue in AMPL function evaluation; Jacobian constraint scaling not applied.")
+                halt_on_ampl_error = self.options.get("halt_on_ampl_error", "yes")
+                if halt_on_ampl_error == "no" :
+                    print("ipopt-watertap: halt_on_ampl_error=no, so continuing with optimization.")
+                else:
+                    raise RuntimeError("Error in AMPL evaluation.\n"
+                            "Run ipopt with halt_on_ampl_error=yes and symbolic_solver_labels=True to see the affected function.")
+            else:
+                print("Error in constraint_autoscale_large_jac")
+                raise
+
+        # this creates the NL file, among other things
+        super()._presolve(*args, **kwds)
+
+    def _postsolve(self):
+        if self._reset_needed:
+            self._reset_scaling_factors()
+            # remove our reference to the model
+            del self._model
+        del self._reset_needed
+        return super()._postsolve()
+
+    def _cache_scaling_factors(self):
+        self._scaling_cache = [ (c, get_scaling_factor(c)) for c in
+                self._model.component_data_objects(
+                    pyo.Constraint, active=True, descend_into=True)]
+
+    def _reset_scaling_factors(self):
+        for c, s in self._scaling_cache:
+            if s is None:
+                unset_scaling_factor(c)
+            else:
+                set_scaling_factor(c, s)
+        del self._scaling_cache
+
+    def _get_option(self, option_name, default_value):
+        # NOTE: options get reset to their original value at the end of the
+        #       OptSolver.solve. The options in _presolve (where this is called)
+        #       are already copies of the original, so it is safe to pop them so
+        #       they don't get sent to Ipopt.
+        option_value = self.options.pop(option_name, None)
+        if option_value is None:
+            option_value = default_value
+        else:
+            if self._tee:
+                print(f"ipopt-watertap: {option_name}={option_value}")
+        return option_value
+
+    def _is_user_scaling(self):
+        if "nlp_scaling_method" not in self.options:
+            self.options["nlp_scaling_method"] = "user-scaling"
+        if self.options["nlp_scaling_method"] != "user-scaling":
+            if self._tee:
+                print("The ipopt-watertap solver is designed to be run with user-scaling. "
+                        f"Ipopt with nlp_scaling_method={self.options['nlp_scaling_method']} will be used instead")
+            return False
+        return True
+
+
+## reconfigure IDAES to use the ipopt-watertap solver
+import idaes
+_default_solver_config_value = idaes.cfg.get("default_solver")
+_idaes_default_solver = _default_solver_config_value._default
```

Comment:
How does this get used?

Reply:
This code gets called when the module is imported, which because of the various imports above happens when you first import `watertap`.

Reply:
I suppose this particular variable is not used, however. I guess there's a question here about adding an additional function (which I didn't create) to undo the step done below (resetting the default solver). In any case you could do it "by hand" by accessing this variable.